internal/pkg/agent/application/enroll/enroll.go

```diff
@@ -0,0 +1,574 @@
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/info"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/configuration"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	"github.com/elastic/elastic-agent/internal/pkg/config"
+	"github.com/elastic/elastic-agent/internal/pkg/core/backoff"
+	monitoringConfig "github.com/elastic/elastic-agent/internal/pkg/core/monitoring/config"
+	"github.com/elastic/elastic-agent/internal/pkg/crypto"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	fleetclient "github.com/elastic/elastic-agent/internal/pkg/fleetapi/client"
+	"github.com/elastic/elastic-agent/internal/pkg/release"
+	"github.com/elastic/elastic-agent/internal/pkg/remote"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+const (
+	EnrollBackoffInit = time.Second * 5
+	EnrollBackoffMax  = time.Minute * 10
+
+	maxRetriesstoreAgentInfo       = 5
+	defaultFleetServerHost         = "0.0.0.0"
+	defaultFleetServerPort         = 8220
+	defaultFleetServerInternalHost = "localhost"
+	defaultFleetServerInternalPort = 8221
+	statusPath                     = "/api/status"
+	apiStatusTimeout               = 15 * time.Second
+	backupSuffix                   = ".enroll.bak"
+)
+
+type saver interface {
+	Save(io.Reader) error
+}
+
+func CheckRemote(ctx context.Context, c fleetclient.Sender) error {
+	ctx, cancel := context.WithTimeout(ctx, apiStatusTimeout)
+	defer cancel()
+
+	resp, err := c.Send(ctx, http.MethodGet, "/api/status", nil, nil, nil)
+	if err != nil {
+		return fmt.Errorf("fail to communicate with Fleet Server API client hosts: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("fleet server ping returned a bad status code: %d", resp.StatusCode)
+	}
+
+	// discard body for proper cancellation and connection reuse
+	_, _ = io.Copy(io.Discard, resp.Body)
+	resp.Body.Close()
+
+	return nil
+}
+
+// BackupConfig creates a backup of currently used fleet config
+func BackupConfig() error {
+	configFile := paths.AgentConfigFile()
+	backup := configFile + backupSuffix
+
+	err := copy.Copy(configFile, backup, copy.Options{
+		PermissionControl: copy.AddPermission(0600),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to backup config file %s -> %s: %w", configFile, backup, err)
+	}
+
+	return nil
+}
+
+// RestoreConfig restores from backup if needed and signals restore was performed
+func RestoreConfig() error {
+	configFile := paths.AgentConfigFile()
+	backup := configFile + backupSuffix
+
+	// check backup exists
+	if _, err := os.Stat(backup); os.IsNotExist(err) {
+		return nil
+	}
+
+	if err := file.SafeFileRotate(configFile, backup); err != nil {
+		return fmt.Errorf("failed to safe rotate backup config file: %w", err)
+	}
+
+	return nil
+}
+
+// CleanBackupConfig removes backup config file
+func CleanBackupConfig() error {
```

Comment:
Is there any reason for this function to be in the `enroll` package and exported?  AFAICT, it's only being called from the coordinator, so perhaps it could live as an unexported function in the `coordinator` package?